From 3a71355357b1fa0403e650156d64e3b877596022 Mon Sep 17 00:00:00 2001
From: "song.han" <song.han@amlogic.com>
Date: Wed, 23 Feb 2022 14:20:28 +0800
Subject: [PATCH 12/95] boottime: change jent_mod_init to delaywork [1/1]

PD#SWPL-71744

Problem:
Jitter rng is doing 1024 rounds of tests and causing
long initialization time

Solution:
change jent_mod_int to delaywork to reduce boot time

Verify:
c2_af400

Signed-off-by: song.han <song.han@amlogic.com>
Change-Id: I90e67854b213dd673077c72bb4830264d17f1f88
---
 crypto/jitterentropy-kcapi.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/crypto/jitterentropy-kcapi.c b/crypto/jitterentropy-kcapi.c
index e8a4165a1874..4bf243bf3085 100644
--- a/crypto/jitterentropy-kcapi.c
+++ b/crypto/jitterentropy-kcapi.c
@@ -197,6 +197,7 @@ static struct rng_alg jent_alg = {
 	}
 };
 
+#ifndef CONFIG_AMLOGIC_BOOT_TIME
 static int __init jent_mod_init(void)
 {
 	int ret = 0;
@@ -208,6 +209,32 @@ static int __init jent_mod_init(void)
 	}
 	return crypto_register_rng(&jent_alg);
 }
+#else
+static struct delayed_work jent_work;
+static void __jent_mod_init(struct work_struct *work)
+{
+	int ret = 0;
+
+	ret = jent_entropy_init();
+	if (ret) {
+		pr_info("jitterentropy: Initialization failed with : %d\n", ret);
+		return;
+	}
+	ret = crypto_register_rng(&jent_alg);
+	if (ret) {
+		pr_err("registering jent_alg failed: %d\n", ret);
+		return;
+	}
+}
+
+static int __init jent_mod_init(void)
+{
+	INIT_DELAYED_WORK(&jent_work, __jent_mod_init);
+	schedule_delayed_work(&jent_work, msecs_to_jiffies(1000));
+
+	return 0;
+}
+#endif
 
 static void __exit jent_mod_exit(void)
 {
-- 
2.25.1

